### ISSUE TITLE
Filmic tonemapping needs tweaking

### ISSUE THREAD

**Issue by smcameron** [smcameron/space-nerds-in-space#273]
I [implemented](https://github.com/smcameron/space-nerds-in-space/commit/865f78b4f61d141d5173905c416f46e9ef1991c8) "filmic tonemapping", [cribbed from oolite](http://www.aegidian.org/bb/viewtopic.php?p=265300#p265300).  But it doesn't look as good (or as subtle) as it does with oolite.

There's a magic formula it uses (from graph_dev_opengl.c):

```
/*
 * Filmic tonemapping cribbed from oolite:
 *
 *      gamma correction
 *      using Jim Hejl's filmic tonemapping and gamma correction approximation.
 *      Normally this would require HDR, but I think it works extremely well in Oolite.
 *      Formula taken from https://www.gdcvault.com/play/1012351/Uncharted-2-HDR
 *      jump to 27:40 in the video. Note the pow 1.0/2.2 is baked into these numbers
 *
 * Perhaps that it normally requires HDR is the reason it doesn't seem to look so
 * great in SNIS.
 */
#define FILMIC_TONEMAPPING \
        "uniform float u_FilmicTonemapping;\n" \
        "vec4 filmic_tonemap(vec4 color) {\n" \
        "       float dont_tonemap = 1.0 - u_FilmicTonemapping;\n" \
        "       vec4 x = max(vec4(0.0), color - 0.004);\n" \
        "       x = (x * (6.2 * x + 0.5)) / (x * (6.2 * x + 1.7) + 0.06);\n" \
        "       return dont_tonemap * color + u_FilmicTonemapping * x;\n" \
        "}\n\n"
```

Here's what it looks like with tonemapping enabled (via renderer menu accessed via hitting SysRq key twice):
![with-filmic-tonemapping](https://user-images.githubusercontent.com/23912/77211360-2c7c9b80-6ada-11ea-8414-4878f569d870.jpg)

Here's what it looks like without tonemapping:

![without-filmic-tonemapping](https://user-images.githubusercontent.com/23912/77211380-3900f400-6ada-11ea-8955-31e49bdb862d.jpg)

It's really slamming the highlights against whiteness.

Adjusting the atmosphere_brightness and the ambient_light tweakable variables helps some, but it seems like it needs to be made more gentle somehow.  Or maybe my lighting model is fubar, and this is just making it easy to see how fubar.  The filmic tonemapping does make the terminator look nice and sharp though.

**Comment by smcameron** [comment]
Ok, I got curious about what this magic function was doing, so I went over to wolframalpha and plotted it, expecting it to have a portion between 0 and 1 that looks like a sigmoid function... but it doesn't.

![weird-function](https://user-images.githubusercontent.com/23912/77215920-42478c00-6aed-11ea-8c44-f85a5ee17da2.jpg)

So then I started playing around with the sigmoid function 1 / (e^-x + 1) and messed around with it until I got the sigmoid shape into the region 0 to 1, and got the steepness what I imagined might be right.

![sigmoid](https://user-images.githubusercontent.com/23912/77215946-702cd080-6aed-11ea-851d-a9440d906a5f.jpg)

Plugged that into my filmic_tonemap function...

```
#define FILMIC_TONEMAPPING \
        "uniform float u_FilmicTonemapping;\n" \
        "vec4 filmic_tonemap(vec4 color) {\n" \
        "       float dont_tonemap = 1.0 - u_FilmicTonemapping;\n" \
        "       // vec4 x = max(vec4(0.0), color - 0.004);\n" \
        "       // x = (x * (6.2 * x + 0.5)) / (x * (6.2 * x + 1.7) + 0.06);\n" \
        "       vec4 x =  max(vec4(0.0), color - 0.004);\n" \
        "       vec4 e = vec4(2.718281828);\n" \
        "       x = (1.0 / (pow(e, -(5.0 * ( 2.0 * x - 1))) + 1));\n" \
        "       return dont_tonemap * color + u_FilmicTonemapping * x;\n" \
        "}\n\n"
```

And got results like this (much more what I thought it'd be like, though more extreme than I was wanting):

![sigmoid-surprise](https://user-images.githubusercontent.com/23912/77215979-a2d6c900-6aed-11ea-947a-f73d7a0310c0.jpg)

![sigmoid-surprise2](https://user-images.githubusercontent.com/23912/77215986-ac603100-6aed-11ea-9622-183eabf90fa8.jpg)

![sigmoid-surprise3](https://user-images.githubusercontent.com/23912/77215993-b2561200-6aed-11ea-8852-c375f337fd74.jpg)

With some more tweaking, I could probably tame that sigmoid to get good results.  However, doing four pow() operations per pixel is probably not too cool (and they happen even if I turn off the filmic tonemapping they way I've done it.)

Maybe tweak it until it looks good, then replace with a lookup table to get the same shape.

**Comment by smcameron** [comment]
Also, the above tonemapping code is modifying the alpha channel.  Probably shouldn't.

**Comment by smcameron** [comment]
Looking at the original magic function,

```
(x * (6.2 * x + 0.5)) / (x * (6.2 * x + 1.7) + 0.06)
```

it's 0 at x=0, 0.8417 at x=1.0, and asymptotically approaches 1.0 as x -> infinity.

![weird-fn](https://user-images.githubusercontent.com/23912/77218367-09b1ad80-6b01-11ea-8d4f-e4f0e9b7237b.jpg)  
![weird-fn-2](https://user-images.githubusercontent.com/23912/77218373-0f0ef800-6b01-11ea-83d5-a257075a55e6.jpg)

It starts off quite steep at x=0, and tapers off as x -> 1.

So this has the effect of lightening dark areas a lot, and squashing bright areas towards a gray of 0.8417 % white.  So the whites will be gray but squashed lighter, and the darks will be gray.  Hmm.

It's not a sigmoid shape.  In that video https://www.gdcvault.com/play/1012351/Uncharted-2-HDR at 22:20 or so he shows a sigmoid graph for film response to light.  So I am confused, as this magic formula is completely missing the bottom part of the sigmoid (27:14 in the video also mentions the "s-shaped curve" immediately before mentioning this magic formula -- which is not S-shaped.).  So I am confused.

**Comment by smcameron** [comment]
It occurred to me that what this "tonemapping" is doing is more or less precisely what the gimp's color curves are doing.

The "magic formula" is essentially doing this, which you would pretty much never do to a photo:

![tonemapping-gone-bad](https://user-images.githubusercontent.com/23912/77226208-f0365300-6b4c-11ea-867e-abcbaae3dba7.jpg)

Perhaps if my data were in linear colorspace rather than sRGB before doing the operation?  (But now I am using words I do not completely understand.)

My naive sigmoid function is also something you'd never do to a photo:

![naive-sigmoid](https://user-images.githubusercontent.com/23912/77226196-e01e7380-6b4c-11ea-95b1-254ae32c3ec6.jpg)

But here's something that you might do to a photo. You might squash the darkest darks even darker, raise up the mid darks a bit, and compress the highlights a bit:

![better-tonemapping](https://user-images.githubusercontent.com/23912/77226227-0f34e500-6b4d-11ea-99d2-8cdcdb0d4b17.jpg)

It occurs to me that I could apply such a curve to a 1x256 gradient from black to white and in this way create a 1x256 texture that could be used as a lookup table.  However, there wouldn't be any knobs.  I could re-implement the gimp's color curves, dynamically modify the lookup table and apply in the shaders (but this would be overkill, and I'm not doing that.)

**Comment by smcameron** [comment]
Reading a bit more...
 
[https://www.kinematicsoup.com/news/2016/6/15/gamma-and-linear-space-what-they-are-how-they-differ](Linear color space)

In particular [this diagram](https://images.squarespace-cdn.com/content/v1/525ed7ade4b0924d2f499223/1465860836325-6SQ4IW87E0HOL034XKYY/ke17ZwdGBToddI8pDm48kLU4AbOIWC21tBEkYh8cvZ8UqsxRUqqbr1mOJYKfIPR7wHSwpYRrzxWb3AxhO8KtM3w1xqCMUWxv9-SrPz8gGiRCRW4BPu10St3TBAUQYVKcz0uF93tz5-I1aD3tEcG_bz2l4dKCG96C-iw77Lt99B91e-d_6cWMNR8_CdnyZyYo/An+image+comparing+gamma+and+linear+pipelines?format=1000w).

Notice the gamma correction at the end of the linear color space pipeline.

[Gamma correction](https://learnopengl.com/Advanced-Lighting/Gamma-Correction) looks a lot like
the "magic formula". See [this diagram](https://learnopengl.com/img/advanced-lighting/gamma_correction_gamma_curves.png) which makes sense because it claimed to have the "1/2.2" built-in, which that is the gamma correction.  So it is apparently supposed to be gamma-correction plus some sort of "filmic tonemapping" in one approximation formula.

Of course all my textures are just made up by my, by eye, so they are going to be in "gamma space".
Then I do lighting calculations on them producing slightly "wrong" results.  I should be removing the gamma correction from the textures, doing the lighting calculations in linear colorspace, then re-applying the gamma correction at the end (i.e. via the magic formula, or another way).

Hmm.  Is this worth it?

**Comment by smcameron** [comment]
Ok, I took a stab at converting to use linear colorspace internally (I think.)

This involves doing using GL_SRGB8 or GL_SRGB8_ALPHA8 (rather than GL_RGB8 or GL_RGBA8) in calls to glTeximage2D() which makes opengl convert from sRGB data in the file...

Then, when I turn on the filmic tonemapping with the magic formula (and also turning ambient_light down to 0.023 (arbitrary) and atmosphere_brightness down to 0.3, I get this:

![gamma+filmic](https://user-images.githubusercontent.com/23912/77231198-cfcbc000-6b6f-11ea-958a-3804d4716ee9.jpg)

Which is not bad.  The terminator is much sharper.  The stars are not tone mapped though.

However, I still have some sort of bug in my cubemap texture (re)loading...

![broken-cubemap](https://user-images.githubusercontent.com/23912/77231225-ff7ac800-6b6f-11ea-8b72-59f54b055c17.jpg)

**Comment by smcameron** [comment]
The bug in the cubemap texture reloading turned out to be I was loading my normalmap textures as sRGB instead of linear so it messes up the lighting.

**Comment by smcameron** [comment]
Alright I think I've about got most of the kinks worked out of the linear pipeline.  Unfortunately you can't dynamically switch between them, because to so so involves reloading all the textures with a different conversion.

I ran two clients, one with the old code and one with the new code.  The old is on the left, the new is on the right.

![new-vs-old](https://user-images.githubusercontent.com/23912/77234544-ad44a180-6b85-11ea-8acd-fa7385c29cdf.jpg)

![new-vs-old2](https://user-images.githubusercontent.com/23912/77234551-b3d31900-6b85-11ea-8461-1e0449167be1.jpg)

![new-vs-old3](https://user-images.githubusercontent.com/23912/77234556-ba619080-6b85-11ea-83b3-b943a77f5dc5.jpg)

It seems to knock the blue out of the atmosphere.  May need to work on the atmosphere shader.  The terminator is a lot sharper in the new version.

**Comment by smcameron** [comment]
* d6a6d31f6a955760e560c9158f3f5e79dc54527b  Implement gamma correction.
* 9912b1c24773c99755656c914bcc27158421c753  Tone down atmosphere shader for tonemapping 
* 721c03731ecc0e189a5484d17fa536bab0c6e0dc  Make green-burst fit better with tonemapping 

Now all textures except normalmaps are treated as sRGB and converted to linear colorspace, and then gamma correction is done in the shader along with the tonemapping via the magic formula, and tonemapping is on by default. (Turning it off, you see the un-gamma-corrected sRGB textures.)

**Comment by MCMic** [comment]
What is the goal of all this?

**Comment by smcameron** [comment]
> What is the goal of all this?

@MCMic Ha!  I don't know.  I fell down a color rabbit hole.  I saw the filmic tonemapping in Oolite's shaders, and I thought "hey, that's just a couple lines to slam onto gl_FragColor,  so easy, 10 minute job, tops." (famous last words) "so let's try it and see how it looks."  And two days later, here we are.  Well, I learned a bunch of stuff I didn't previously know.

But anyway, if I ever want to/am able to get to physically based rendering with proper materials, etc. I need to be operating with a linear color space in the shaders anyway.  And it does solve one thing that has bugged me for a long, long time.  I had noticed that the transition from light to dark across the terminator of my planets seemed softer than what you'd often see in, e.g. [NASA photos](https://www.nasa.gov/image-feature/jpl/pia21345/so-far-from-home) and this color correction business seems to solve that.  And the gas giants (to me) look better... not so sure about the earthlike planets.  (OTOH, maybe the gas giants aren't better, just different, and maybe I'm just changing things to be changing things while I hide out from Corona virus.  I *think* they look better... for now anyway.)
